src/copy.rs

```diff
@@ -0,0 +1,263 @@
+    W: AsyncWrite + Unpin + ?Sized,
+{
+    type Output = std::io::Result<u64>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            let me = &mut *self;
+            let buffer = ready!(Pin::new(&mut *me.reader).poll_fill_buf(cx))?;
+            if buffer.is_empty() {
+                ready!(Pin::new(&mut self.writer).poll_flush(cx))?;
+                return Poll::Ready(Ok(self.amt));
+            }
+
+            let i = ready!(Pin::new(&mut *me.writer).poll_write(cx, buffer))?;
+            if i == 0 {
+                return Poll::Ready(Err(std::io::ErrorKind::WriteZero.into()));
+            }
+            if me.send {
+                me.metrics.increment_send(i as u64);
+            } else {
+                me.metrics.increment_recv(i as u64);
+            }
+            let old = self.amt;
+            self.amt += i as u64;
+
+            // If we were below the resize threshold before but are now above it, trigger the buffer to resize
+            if old < RESIZE_THRESHOLD && RESIZE_THRESHOLD <= self.amt {
+                Pin::new(&mut *self.reader).resize();
+            }
+            Pin::new(&mut *self.reader).consume(i);
+        }
+    }
+}
+
+// BufReader is a fork of Tokio's type with resize support
+pin_project! {
+    pub struct BufReader<R> {
+        #[pin]
+        inner: R,
+        buf: Box<[u8]>,
+        pos: usize,
+        cap: usize,
+    }
+}
+
+impl<R: AsyncRead> BufReader<R> {
+    /// Creates a new `BufReader` with a default buffer capacity. The default is currently INITIAL_BUFFER_SIZE
+    pub fn new(inner: R) -> Self {
+        let buffer = vec![0; INITIAL_BUFFER_SIZE];
+        Self {
+            inner,
+            buf: buffer.into_boxed_slice(),
+            pos: 0,
+            cap: 0,
+        }
+    }
+
+    fn get_ref(&self) -> &R {
+        &self.inner
+    }
+
+    fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut R> {
+        self.project().inner
+    }
+}
+
+impl<R: AsyncRead> ResizeBufRead for BufReader<R> {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&[u8]>> {
+        let me = self.project();
+
+        // If we've reached the end of our internal buffer then we need to fetch
+        // some more data from the underlying reader.
+        // Branch using `>=` instead of the more correct `==`
+        // to tell the compiler that the pos..cap slice is always valid.
+        if *me.pos >= *me.cap {
+            debug_assert!(*me.pos == *me.cap);
+            let mut buf = tokio::io::ReadBuf::new(me.buf);
+            ready!(me.inner.poll_read(cx, &mut buf))?;
+            *me.cap = buf.filled().len();
+            *me.pos = 0;
+        }
+        Poll::Ready(Ok(&me.buf[*me.pos..*me.cap]))
+    }
+
+    fn consume(self: Pin<&mut Self>, amt: usize) {
+        let me = self.project();
+        *me.pos = cmp::min(*me.pos + amt, *me.cap);
+    }
+
+    fn resize(self: Pin<&mut Self>) {
```

Comment:
Do we have any info on how expensive resizes are in practice?
Is it something we could measure? There's clearly a tradeoff but the buffer usage is only part of it.

Reply:
Cost to do 1 malloc (16k) + 1 memcopy (up to 1k). Its basically not noticeable -- its slow to do lots of memcpy, but 1 per connection is pretty tiny.

I did actually try to measure it by seeing if I saw a spike in latency when the resize kicked in; I wasn't able to detect one which gave me the impression it is negligible in the grand scheme of the request flow